### PR TITLE
feat(core): add api locking interface POC

### PR DESCRIPTION
**Pull request by Abhicodes-crypto**
WIP

## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description

This PR contains the functions used for API locking. 

### Additional Changes

- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

Api locking is required to prevent threads from modifying the same resource simultaneously. 

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/core/locking.rs [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+pub enum ActionOnWaitTimeout {}
+
+pub struct LockingInput {
+    pub unique_locking_key: String,
+    pub api_identifier: String,
+    pub action_on_wait_timeout: ActionOnWaitTimeout,
+    pub merchant_id: String,
+}
+
+pub async fn get_key_and_lock_resource(
```
Comment (jarnura) [inline crates/router/src/core/locking.rs:28]: Change the function name, `get_key` operation is not happening inside the function.

And if this function only going to works on a key type, can't we have a unit struct for each type.

And `Trait` `LockResource` having a `lock()` method

File: crates/router/src/core/locking.rs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    locking_input: LockingInput,
+    _should_retry: bool,
+) -> RouterResult<LockStatus> {
+    let api_identifier = locking_input.api_identifier.as_str();
+
+    let is_locking_enabled_for_merchant = true;
+    let is_locking_enabled_on_api = true;
+
+    // let get_expiry_time = get_expiry_time_from_redis_based_on_connector_pmd_pm();
+    if is_locking_enabled_for_merchant && is_locking_enabled_on_api {
```
Comment (jarnura) [inline crates/router/src/core/locking.rs:39]: We can have a discuss why redis? and why not environment variables

File: crates/router/src/core/locking.rs [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+pub async fn lock_resource(
+    state: &routes::AppState,
+    locking_key: String,
+    expiry_in_seconds: u64,
+    retries: i32,
+    delay_between_retries_in_seconds: i64,
+    _api_identifier: &str,
+) -> RouterResult<LockStatus> {
+    let redis_key_for_lock = get_redis_key_for_locks(locking_key);
```
Comment (jarnura) [inline crates/router/src/core/locking.rs:71]: `get_redis_key_for_locks` are we doing mulitple locks?

File: crates/router/src/core/locking.rs [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    let redis_value_for_lock = true; // should get session id or request_id as we need info of who acquired the lock.
+    acquire_lock_on_resource_in_redis(
+        state,
+        redis_key_for_lock.as_str(),
+        redis_value_for_lock,
+        expiry_in_seconds,
+        delay_between_retries_in_seconds
```
Comment (jarnura) [inline crates/router/src/core/locking.rs:78]: time value are parsed in lots of places, parse once and parse in single place. Avoid parsing multiple times.

**Review by jarnura — CHANGES_REQUESTED** [review]

**Comment by shakthi-22** [comment]
This is being worked on again by @Abhicodes-crypto

**Pull request closed without merging**